test-suite/README.new.md

```diff
@@ -0,0 +1,324 @@
+diff. You can use `make promote` in order to promote all the actual output files
+form the `_build` directory to be the expected output.
+
+ To have finer control, use `dune promote` on a specific target such as
+ ```sh
+ dune promote test-suite/bugs/mybug.out
+ ```
+
+### Output Test Example
+
+Adding `_test.v` to `output/` for instance will yield the following when running
+`make test`:
+```sh
+Error: No rule found for test-suite/output/_test.out
+-> required by alias test-suite/output/runtest in
+   test-suite/test_suite_rules.sexp:56694
+make: *** [Makefile:122: test] Error 1
+```
+To fix this, we `touch` the file:
+```sh
+touch test-suite/output/_test.out
+```
+and rerun `make test`. Now Dune will complain that the output is different:
+```diff
+File "test-suite/output/_test.out", line 1, characters 0-0:
+diff --git a/_build/default/test-suite/output/_test.out b/_build/default/test-suite/output/_test.v.log
+index e69de29bb2..a7de56ea32 100644
+--- a/_build/default/test-suite/output/_test.out
++++ b/_build/default/test-suite/output/_test.v.log
+@@ -0,0 +1,2 @@
++Type : Type
++     : Type
+make: *** [Makefile:122: test] Error 1
+```
+Running `make promote` will now update the `.out` file with the new output:
+```sh
+Promoting _build/default/test-suite/output/_test.v.log to
+  test-suite/output/_test.out.
+```
+Finally when running `make test`, Dune will report no problems.
+
+## Running Single Tests
+
+All tests output targets such as .vo files or .log files. You can tell Dune to
+build these targets directly. For example:
+
+```sh
+dune build test-suite/bugs/bug_1446.vo
+```
+
+However this is not ideal, since it requires you to know exactly where the test
+is located and at least one of the targets it is expected to produce.
+
+To remedy this, the test-suite will provide a *Dune alias* for each of the `.v`
+file tests. Simply write:
+
+```sh
+dune build @bug_1446
+```
+
+to run the test directly. Note the dropping of the ".v" file suffix. You can of
+course build multiple aliases too.
+
+Currently only the `.v` file tests have support for aliases. Most of the other
+shell tests do not yet have support. There is no reason for this, it is just a
+matter of adding the alias generation where needed. See the generation for misc
+tests for an example of aliases on non-v file tests.
+
+## Running Multiple Tests
+
+You can run all the tests appearing in a specific directory if you build the
+directory directly. For example, to run all the tests in `test-suite/bugs` you
+can do:
+
+```sh
+dune build test-suite/bugs
+```
+
+This will build all the targets appearing in that directory. As stated above,
+you can build multiple targets by listing them out in `dune build`:
+
+```sh
+dune build @test1 @test2 test-suite/some-other-tests
+```
+
+## Running Single Tests Without Rebuilding Deps
+
+You can run a test without compiling its dependencies simply by running:
+
+```sh
+dune build @DiscrR.replay
```

Comment:
This errors on the coqchk step for me
~~~
Fatal Error: User error: The file DiscrR__replay_0c4a9caccfc716c5b3fd8a3828cdafef.vo contains library
              DiscrR and not library
              DiscrR__replay_0c4a9caccfc716c5b3fd8a3828cdafef
~~~

Reply:
Also it seems the replay aliases don't do the output check for output tests.

Reply:
I don't currently have support for replaying coqchk. The fact that coqchk has to have the same name for the vo makes doing this difficult.

Reply:
also also the replay targets aren't compatible with tests where the name of the file matters, eg univbinders:
~~~
File "./UnivBinders__replay_ff11647412b7cd297f02f3631133cf59.v", line 64, characters 0-49:
Error: Undeclared universe UnivBinders.mono.uu.
~~~
I don't remember if coqc allows -topfile, alternatively maybe the replay files could be in a subdirectory .replay/ instead of using some hash suffix.